src/analysis/asymptotics.lean

```diff
@@ -0,0 +1,438 @@
+end
+
+section
+variables [has_norm β] [normed_field γ]
+
+theorem is_bigo_mul_right {f : α → β} {g : α → γ} {l : filter α} {c : γ} (hc : c ≠ 0) :
+  is_bigo f (λ x, c * g x) l ↔ is_bigo f g l :=
+begin
+  split,
+  { rintro ⟨c', h'⟩,
+    use (c' * ∥c∥),
+    convert h', ext x, dsimp,
+    rw [normed_field.norm_mul, mul_assoc] },
+  have cne0 : ∥c∥ ≠ 0, by rw [ne, norm_eq_zero]; exact hc,
+  rintro ⟨c', h'⟩,
+  use (c' * ∥c∥⁻¹),
+  convert h', ext x, dsimp,
+  rw [normed_field.norm_mul, mul_assoc, ←mul_assoc ∥c∥⁻¹, inv_mul_cancel cne0, one_mul]
+end
+
+theorem is_littleo_mul_right {f : α → β} {g : α → γ} {l : filter α} {c : γ} (hc : c ≠ 0) :
+  is_littleo f (λ x, c * g x) l ↔ is_littleo f g l :=
+begin
+  have cne0 : ∥c∥ ≠ 0, by rw [ne, norm_eq_zero]; exact hc,
+  have cpos : ∥c∥ > 0, from lt_of_le_of_ne (norm_nonneg _) (ne.symm cne0),
+  split,
+  { intros h' c',
+    filter_upwards [h' (∥c∥ * c')], dsimp,
+    intro x, rw [normed_field.norm_mul, mul_assoc],
+    intro h₀, exact le_of_mul_le_mul_left h₀ cpos },
+  intros h' c',
+  convert h' (∥c∥⁻¹ * c'), ext x, dsimp,
+  symmetry,
+  rw [normed_field.norm_mul, mul_assoc, mul_comm],
+  apply div_le_iff' cpos
+end
+
+end
+
+/-
+Note: the method in the next two sections can be used for the integers, or more generally any normed
+ring satisfying `∥a * b∥ = ∥a∥ * ∥b∥`. It is not clear whether it is worth stating it in that
+generality.
+-/
+
+section
+variables {k : Type*} [normed_field k] [normed_space k β] [has_norm γ]
+
+theorem is_bigo_smul_left {f : α → β} {g : α → γ} {l : filter α} {c : k} (hc : c ≠ 0) :
+  is_bigo (λ x, c • f x) g l ↔ is_bigo f g l :=
+begin
+  have cne0 : ∥c∥ ≠ 0, by rw [ne, norm_eq_zero]; exact hc,
+  rw [←is_bigo_norm_left], simp only [norm_smul],
+  rw [is_bigo_mul_left cne0, is_bigo_norm_left]
+end
+
+theorem is_littleo_smul_left {f : α → β} {g : α → γ} {l : filter α} {c : k} (hc : c ≠ 0) :
+  is_littleo (λ x, c • f x) g l ↔ is_littleo f g l :=
+begin
+  have cne0 : ∥c∥ ≠ 0, by rw [ne, norm_eq_zero]; exact hc,
+  rw [←is_littleo_norm_left], simp only [norm_smul],
+  rw [is_littleo_mul_left cne0, is_littleo_norm_left]
+end
+
+end
+
+section
+variables {k : Type*} [has_norm β] [normed_field k] [normed_space k γ]
+
+theorem is_bigo_smul_right {f : α → β} {g : α → γ} {l : filter α} {c : k} (hc : c ≠ 0) :
+  is_bigo f (λ x, c • g x) l ↔ is_bigo f g l :=
+begin
+  have cne0 : ∥c∥ ≠ 0, by rw [ne, norm_eq_zero]; exact hc,
+  rw [←is_bigo_norm_right], simp only [norm_smul],
+  rw [is_bigo_mul_right cne0, is_bigo_norm_right]
+end
+
+theorem is_littleo_smul_right {f : α → β} {g : α → γ} {l : filter α} {c : k} (hc : c ≠ 0) :
+  is_littleo f (λ x, c • g x) l ↔ is_littleo f g l :=
+begin
+  have cne0 : ∥c∥ ≠ 0, by rw [ne, norm_eq_zero]; exact hc,
+  rw [←is_littleo_norm_right], simp only [norm_smul],
+  rw [is_littleo_mul_right cne0, is_littleo_norm_right]
+end
+
+end
+
+/-
+"transitivity" rules
+-/
+
+theorem is_bigo_of_is_bigo_of_is_bigo [has_norm β] [normed_group γ] [has_norm δ]
```

Comment:
These names look a little unwieldy. How about `is_bigo.trans` for this and the little o version, and `is_littleo.trans_bigo` and `is_bigo.trans_littleo` for the other two? (I admit that they are adhering to the naming convention, though, so maybe an alias is in order, if people actually guess this name.)